File: 739-daily-temperatures/739-daily-temperatures.scala
import scala.collection.mutable._

object Solution {
    case class Temperature(val index: Int, val value: Int)

    def dailyTemperatures(temperatures: Array[Int]): Array[Int] = {
        val res: Array[Int] = Array.fill(temperatures.length)(0)
        val stack: Stack[Temperature] = Stack()

        for(i <- 0 to temperatures.length - 1) {
            val idx = temperatures.length - 1 - i
            val temp = temperatures(idx)

            while(stack.nonEmpty && stack.top.value <= temp) stack.pop
            res(idx) = if (stack.isEmpty) 0 else (stack.top.index - idx)
            stack.push(Temperature(idx, temp))
        }

        res
    }
}

File: 739-daily-temperatures/739-daily-temperatures.java
class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Deque<int[]> stack = new LinkedList<>();
        int[] res = new int[temperatures.length];
        
        for(int i = temperatures.length - 1; i >= 0; i--) {
            while(!stack.isEmpty() && stack.peek()[0] <= temperatures[i]) stack.pop();
            res[i] = (stack.isEmpty())? 0 : stack.peek()[1] - i;
            stack.push(new int[] { temperatures[i], i });
        }
        
        return res;
    }
}